bip-0331.mediawiki

```diff
@@ -0,0 +1,394 @@
+Service attack vectors, and policy limitations that could be leveraged to censor transactions (aka
+"pinning attacks").  This protocol is extensible to support multiple types of packages based on
+future desired use cases.</ref>
+<ref>'''Why send one message per version instead of sending a single message for the highest
+supported version?''' It should also be possible to support some subset of the existing package
+types. For example, if a node's mempool policy doesn't support or a node implementation deprecates
+version 2, it should be easy to announce support for version 1, but not version 2, packages.</ref>
+
+[[File:./bip-0331/version_negotiation.png|600px]]
+
+===Combined Hash===
+
+A "combined hash" serves as a unique "package id" for some list of unique transactions.
+
+The combined hash of a package of transactions is equal to the hash of each transaction's wtxid, concatenated in lexicographical order.
+
+The combined hash can be calculated using just the wtxids of the transactions.
+
+===New Messages===
+
+Four new protocol messages and two inv types are added.
+
+====sendpackages====
+
+{|
+|  Field Name  ||  Type  ||  Size  ||  Purpose
+|-
+|version || uint32_t || 4 || Denotes a package version supported by the node.
+|-
+|}
+
+# The "sendpackages" message has the structure defined above, with pchCommand == "sendpackages".
+
+# During version handshake, nodes should send a "sendpackages" message to indicate they support package relay. Nodes should send a "sendpackages" message for each version they support.
+
+# Upon receipt of a "sendpackages" message with a version that is not supported, a node must treat the peer as if it never received the message.
+
+# The "sendpackages" message MUST be sent before sending a "verack" message. If a "sendpackages" message is received afer "verack", the sender may be disconnected.
+
+# A node that sends "sendpackages" MUST also send "wtxidrelay". Upon receipt of a "verack", if the sender has sent a "sendpackages" but not "wtxidrelay", the sender may be disconnected.
+
+# Upon successful connection ("verack" sent by both peers), a node may relay packages with the peer if they did not set "fRelay" to false in the "version" message, both peers sent "wtxidrelay", and both peers sent "sendpackages" for matching version(s). Peers should relay packages corresponding to versions that both sent "sendpackages" for.<ref>'''Is it ok to send "sendpackages" to a peer that specified fRelay=false in their "version" message?'''
+Yes, this is allowed in order to reduce the number of negotiation steps. This means nodes can
+announce features without first checking what the other peer has sent, and then apply logic at the
+end. See [https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-May/020510.html discussion].
+</ref>
+
+# Version=0 packages correspond to ancestor packages, i.e. "ancpkginfo" and "MSG_ANCPKGINFO".
```

Comment:
I think it'd be better to pull this definition out and make it more explicit. It'd also be useful to distinguish between the parts of this proposal that are specific to version 0 (`MSG_ANCPKGINFO` and `ancpkginfo`) and the parts that are generic to any kind of package (`sendpackages`, `getpkgtxns`, `pkgtxns` and `MSG_PKGTXNS`)

Reply:
Done